### PR TITLE
Reorder structure: toolchain-id before config-id

### PR DESCRIPTION
**Pull request by NeroBurner**
based on https://github.com/ruslo/hunter/pull/937 for modeline support
fixes: https://github.com/ruslo/hunter/issues/931

- call hunter_calculate_toolchain_sha1 before
  hunter_calculate_config_sha1
- switch config and toolchain SHA1 in folder structure to be
  _BASE/${HUNTER_ID}/${HUNTER_TOOLCHAIN_ID}/${HUNTER_CONIG_ID}
- update modules which create files in HUNTER_TOOLCHAIN_ID_PATH to use
  config path instead

### CODE REVIEW

**Comment by NeroBurner** [comment]
I got an error with cache or cellar. Tested with `examples/Eigen`. The build and install steps work fine, but when doing caching the following error occurs

```
-- [hunter *** DEBUG *** 2017-08-07T09:44:20] Unpacked successfully
-- [hunter *** DEBUG *** 2017-08-07T09:44:20] Creating list of files and directories
-- [hunter *** DEBUG *** 2017-08-07T09:44:20] Creating directories
-- [hunter *** DEBUG *** 2017-08-07T09:44:20] Removing old files
-- [hunter *** DEBUG *** 2017-08-07T09:44:20] Linking files
-- [hunter *** DEBUG *** 2017-08-07T09:44:20] [/home/gc/repos/hunter/hunter/_Base/Cellar/2a54e1fce126720c44bc848c9bf4a2365c5efac6/2a54e1f] > "/usr/bin/cmake" "-DHUNTER_INSTALL_PREFIX=/home/gc/repos/hunter/hunter/_Base/xxxxxxx/e1266bb/fb1da22/Install" "-DLIST_OF_FILES=/home/gc/repos/hunter/hunter/_Base/Cellar/2a54e1fce126720c44bc848c9bf4a2365c5efac6/2a54e1f/files.list" "-DSHELL_LINK_SCRIPT=/home/gc/repos/hunter/hunter/_Base/Cellar/2a54e1fce126720c44bc848c9bf4a2365c5efac6/2a54e1f/link-all.sh" "-DCELLAR_RAW_DIRECTORY=/home/gc/repos/hunter/hunter/_Base/Cellar/2a54e1fce126720c44bc848c9bf4a2365c5efac6/2a54e1f/raw" "-DPYTHON_LINK_SCRIPT=/home/gc/repos/hunter/hunter/scripts/link-all.py" "-P" "/home/gc/repos/hunter/hunter/scripts/link-all.cmake"

[hunter ** INTERNAL **] Link script failed: 1, -- Found PythonInterp: /usr/bin/python3 (found suitable version "3.5.2", minimum required is "3") 
, Link files using Python: /usr/bin/python3
CMake Error at /home/gc/repos/hunter/hunter/scripts/link-all.cmake:69 (message):
  Python script failed:
  /usr/bin/python3
[hunter ** INTERNAL **] /home/gc/repos/hunter/hunter/scripts/link-all.py
[hunter ** INTERNAL **] --list
[hunter ** INTERNAL **] /home/gc/repos/hunter/hunter/_Base/Cellar/2a54e1fce126720c44bc848c9bf4a2365c5efac6/2a54e1f/files.list
[hunter ** INTERNAL **] --cellar
[hunter ** INTERNAL **] /home/gc/repos/hunter/hunter/_Base/Cellar/2a54e1fce126720c44bc848c9bf4a2365c5efac6/2a54e1f/raw
[hunter ** INTERNAL **] --dest
[hunter ** INTERNAL **] /home/gc/repos/hunter/hunter/_Base/xxxxxxx/e1266bb/fb1da22/Install,
  1, , Traceback (most recent call last):

    File "/home/gc/repos/hunter/hunter/scripts/link-all.py", line 52, in <module>
      run_link()
    File "/home/gc/repos/hunter/hunter/scripts/link-all.py", line 43, in run_link
      pool = multiprocessing.Pool(processes=proc_num)
    File "/usr/lib/python3.5/multiprocessing/context.py", line 118, in Pool
      context=self.get_context())
    File "/usr/lib/python3.5/multiprocessing/pool.py", line 150, in __init__
      self._setup_queues()
    File "/usr/lib/python3.5/multiprocessing/pool.py", line 243, in _setup_queues
      self._inqueue = self._ctx.SimpleQueue()
    File "/usr/lib/python3.5/multiprocessing/context.py", line 111, in SimpleQueue
      return SimpleQueue(ctx=self.get_context())
    File "/usr/lib/python3.5/multiprocessing/queues.py", line 324, in __init__
      self._rlock = ctx.Lock()
    File "/usr/lib/python3.5/multiprocessing/context.py", line 66, in Lock
      return Lock(ctx=self.get_context())
    File "/usr/lib/python3.5/multiprocessing/synchronize.py", line 163, in __init__
      SemLock.__init__(self, SEMAPHORE, 1, 1, ctx=ctx)
    File "/usr/lib/python3.5/multiprocessing/synchronize.py", line 60, in __init__
      unlink_now)

  PermissionError: [Errno 13] Permission denied




[hunter ** INTERNAL **] [Directory:/home/gc/repos/hunter/hunter/cmake/projects/Eigen]

------------------------------ WIKI -------------------------------
    https://github.com/ruslo/hunter/wiki/error.internal
-------------------------------------------------------------------

CMake Error at /home/gc/repos/hunter/hunter/cmake/modules/hunter_wiki.cmake:12 (message):
Call Stack (most recent call first):
  /home/gc/repos/hunter/hunter/cmake/modules/hunter_internal_error.cmake:13 (hunter_wiki)
  /home/gc/repos/hunter/hunter/cmake/modules/hunter_unpack_directory.cmake:153 (hunter_internal_error)
  /home/gc/repos/hunter/hunter/cmake/modules/hunter_save_to_cache.cmake:103 (hunter_unpack_directory)
  /home/gc/repos/hunter/hunter/cmake/modules/hunter_download.cmake:608 (hunter_save_to_cache)
  /home/gc/repos/hunter/hunter/cmake/projects/Eigen/hunter.cmake:103 (hunter_download)
  /home/gc/repos/hunter/hunter/cmake/modules/hunter_add_package.cmake:53 (include)
  CMakeLists.txt:12 (hunter_add_package)


-- Configuring incomplete, errors occurred!
See also "/home/gc/repos/hunter/hunter/examples/Eigen/_build/CMakeFiles/CMakeOutput.log".
```

File: cmake/modules/hunter_add_package.cmake [hunk 1]
```diff
@@ -54,3 +54,5 @@ macro(hunter_add_package)
     hunter_status_debug("load: ${x} ... end")
   endforeach()
 endmacro()
+
+# vim:set shiftwidth=2 tabstop=2 softtabstop=2 expandtab:
```
Comment (ruslo) [inline cmake/modules/hunter_add_package.cmake]: Please remove

**Comment by NeroBurner** [comment]
- removed the vim modelines
- update documentation

tree -d of _Base directory with installed headers omitted
```
_Base/
|-- Cache
|   |-- meta
|   |   `-- e1266bb
|   |       `-- Eigen
|   |           `-- 3.3.4-p0
|   |               `-- 49dee30
|   |                   `-- 4cd415e
|   |                       `-- a49b0e5
|   |                           `-- da39a3e
|   `-- raw
|-- Cellar
|   `-- cddcb40101ee3818875d3bce3077517b1205b63c
|       `-- cddcb40
|           `-- raw
|               |-- include
|               |   `-- eigen3
|               |-- licenses
|               |   `-- Eigen
|               `-- share
|                   |-- eigen3
|                   |   `-- cmake
|                   `-- pkgconfig
|-- Download
|   `-- Eigen
|       `-- 3.3.4-p0
|           `-- 49dee30
`-- xxxxxxx
    `-- e1266bb
        `-- bace980
            |-- Build
            |   `-- Eigen
            |       |-- Build
            |       |-- Install
            |       |   |-- include
            |       |   |   `-- eigen3
            |       |   |-- licenses
            |       |   |   `-- Eigen
            |       |   `-- share
            |       |       |-- eigen3
            |       |       |   `-- cmake
            |       |       `-- pkgconfig
            |       `-- Source
            `-- Install
                |-- include
                |   `-- eigen3
                |-- licenses
                |   `-- Eigen
                `-- share
                    |-- eigen3
                    |   `-- cmake
                    `-- pkgconfig
```

**Comment by ruslo** [comment]
May be something wrong with working directory of execute_process, try this version:
* https://github.com/ruslo/hunter/releases/tag/v0.19.60

**Comment by NeroBurner** [comment]
rebased on v0.19.60, still same error

**Comment by ruslo** [comment]
```
  /usr/bin/python3
[hunter ** INTERNAL **] /home/gc/repos/hunter/hunter/scripts/link-all.py
[hunter ** INTERNAL **] --list
[hunter ** INTERNAL **] /home/gc/repos/hunter/hunter/_Base/Cellar/2a54e1fce126720c44bc848c9bf4a2365c5efac6/2a54e1f/files.list
[hunter ** INTERNAL **] --cellar
[hunter ** INTERNAL **] /home/gc/repos/hunter/hunter/_Base/Cellar/2a54e1fce126720c44bc848c9bf4a2365c5efac6/2a54e1f/raw
[hunter ** INTERNAL **] --dest
[hunter ** INTERNAL **] /home/gc/repos/hunter/hunter/_Base/xxxxxxx/e1266bb/fb1da22/Install
```

Can you try to run it manually, like

```
/usr/bin/python3 /home/gc/repos/hunter/hunter/scripts/link-all.py --list /home/gc/repos/hunter/hunter/_Base/Cellar/2a54e1fce126720c44bc848c9bf4a2365c5efac6/2a54e1f/files.list --cellar /home/gc/repos/hunter/hunter/_Base/Cellar/2a54e1fce126720c44bc848c9bf4a2365c5efac6/2a54e1f/raw --dest /home/gc/repos/hunter/hunter/_Base/xxxxxxx/e1266bb/fb1da22/Install
```

Same error?

**Comment by NeroBurner** [comment]
```
/usr/bin/python3 /home/gc/repos/hunter/hunter/scripts/link-all.py  --list /home/gc/repos/hunter/hunter/_Base/Cellar/fc4326be70f5d5cde30f8128c4f3aec5e49a4017/fc4326b/files.list  --cellar /home/gc/repos/hunter/hunter/_Base/Cellar/fc4326be70f5d5cde30f8128c4f3aec5e49a4017/fc4326b/raw  --dest /home/gc/repos/hunter/hunter/_Base/xxxxxxx/e1266bb/bace980/Install 
Traceback (most recent call last):
  File "/home/gc/repos/hunter/hunter/scripts/link-all.py", line 52, in <module>
    run_link()
  File "/home/gc/repos/hunter/hunter/scripts/link-all.py", line 43, in run_link
    pool = multiprocessing.Pool(processes=proc_num)
  File "/usr/lib/python3.5/multiprocessing/context.py", line 118, in Pool
    context=self.get_context())
  File "/usr/lib/python3.5/multiprocessing/pool.py", line 150, in __init__
    self._setup_queues()
  File "/usr/lib/python3.5/multiprocessing/pool.py", line 243, in _setup_queues
    self._inqueue = self._ctx.SimpleQueue()
  File "/usr/lib/python3.5/multiprocessing/context.py", line 111, in SimpleQueue
    return SimpleQueue(ctx=self.get_context())
  File "/usr/lib/python3.5/multiprocessing/queues.py", line 324, in __init__
    self._rlock = ctx.Lock()
  File "/usr/lib/python3.5/multiprocessing/context.py", line 66, in Lock
    return Lock(ctx=self.get_context())
  File "/usr/lib/python3.5/multiprocessing/synchronize.py", line 163, in __init__
    SemLock.__init__(self, SEMAPHORE, 1, 1, ctx=ctx)
  File "/usr/lib/python3.5/multiprocessing/synchronize.py", line 60, in __init__
    unlink_now)
PermissionError: [Errno 13] Permission denied
```

All files and directories exists
```
file /home/gc/repos/hunter/hunter/scripts/link-all.py
/home/gc/repos/hunter/hunter/scripts/link-all.py: Python script, ASCII text executable

file /home/gc/repos/hunter/hunter/_Base/Cellar/fc4326be70f5d5cde30f8128c4f3aec5e49a4017/fc4326b/files.list
/home/gc/repos/hunter/hunter/_Base/Cellar/fc4326be70f5d5cde30f8128c4f3aec5e49a4017/fc4326b/files.list: ASCII text

file /home/gc/repos/hunter/hunter/_Base/Cellar/fc4326be70f5d5cde30f8128c4f3aec5e49a4017/fc4326b/raw 
/home/gc/repos/hunter/hunter/_Base/Cellar/fc4326be70f5d5cde30f8128c4f3aec5e49a4017/fc4326b/raw: directory

file /home/gc/repos/hunter/hunter/_Base/xxxxxxx/e1266bb/bace980/Install
/home/gc/repos/hunter/hunter/_Base/xxxxxxx/e1266bb/bace980/Install: directory
```

**Comment by NeroBurner** [comment]
was a problem with the schroot I was using https://stackoverflow.com/questions/2009278/python-multiprocessing-permission-denied#2009505

solution was to add the following line to `/etc/schroot/default/fstab`
```
none            /dev/shm        tmpfs   rw,nosuid,nodev,noexec 0 0
```

on default `/dev` is mounted into the schroot, but `/dev/shm` is a mount within a mount, which is not propagated

**Comment by ruslo** [comment]
> was a problem with the schroot

Thanks, I've added link to error message:
* https://github.com/ruslo/hunter/commit/be47752e7489809c51c0e4effa1dac12e7c56283

I hope this code will be removed in future, so let it be like this for now.

**Comment by NeroBurner** [comment]
which code will be removed? the python script?

**Comment by ruslo** [comment]
> which code will be removed? the python script?

Yep, https://gitlab.kitware.com/cmake/cmake/issues/16926

**Comment by NeroBurner** [comment]
rebased and squashed

**Comment by NeroBurner** [comment]
fixed hunter tests

**Comment by ruslo** [comment]
Thanks. I need to carefully review it. May take some time.

**Comment by ruslo** [comment]
Follow ups:
* https://github.com/ruslo/hunter/commit/f8de23500af59cbc657092df4648f97b8ef494d6
* https://github.com/ruslo/hunter/commit/0b875a1dc2695d5f1240e946b9ee06ddf5b8114c
* https://github.com/ruslo/hunter/commit/1a2b1c05002e18e251b066d2768d7efa5719f88d

Released:
* https://github.com/ruslo/hunter/releases/tag/v0.19.67

**Pull request merged**